temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -246,9 +252,70 @@ private StartWorkflowExecutionRequest overrideStartWorkflowExecutionRequest(
     if (taskTimeoutMillis != Durations.toMillis(request.getWorkflowTaskTimeout())) {
       request.setWorkflowTaskTimeout(Durations.fromMillis(taskTimeoutMillis));
     }
+
+    SearchAttributes.Builder searchAttributes = request.getSearchAttributes().toBuilder();
+    Map<String, Payload> payloadMap = request.getSearchAttributes().getIndexedFieldsMap();
+    for (Map.Entry<String, Payload> searchAttribute : payloadMap.entrySet()) {
+      String key = searchAttribute.getKey();
+      SearchAttributeType type = searchAttributeToType(key);
+      if (type != null && !SearchAttributeType.Unspecified.name().equals(type.name())) {
+        searchAttributes.putIndexedFields(
+            key,
+            searchAttribute
+                .getValue()
+                .toBuilder()
+                .putMetadata(
+                    EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_JSON)
+                .putMetadata(EncodingKeys.METADATA_TYPE_KEY, ByteString.copyFromUtf8(type.name()))
+                .build());
+      } else {
+        searchAttributes.removeIndexedFields(key);
+        throw Status.INVALID_ARGUMENT.asRuntimeException();
+      }
+    }
+    request.setSearchAttributes(searchAttributes);
     return request.build();
   }
 
+  private SearchAttributeType searchAttributeToType(String key) {
+    SearchAttributesUtil.RegisteredSearchAttributes searchAttribute;
+    try {
+      searchAttribute = SearchAttributesUtil.RegisteredSearchAttributes.valueOf(key);
+    } catch (IllegalArgumentException e) {
+      return null;
+    }
+    switch (searchAttribute) {
+      case BatcherNamespace:
+      case BatcherUser:
+      case BinaryChecksums:
+      case ExecutionStatus:
+      case RunId:
+      case TaskQueue:
+      case TemporalChangeVersion:
+      case WorkflowId:
+      case WorkflowType:
+      case CustomKeywordField:
+      case CustomStringField:
+        return javaTypeToEncodedType(String.class);
+      case CloseTime:
+      case ExecutionTime:
+      case StartTime:
+      case CustomDatetimeField:
+        return javaTypeToEncodedType(LocalDateTime.class);
+      case ExecutionDuration:
+      case HistoryLength:
+      case StateTransitionCount:
+      case CustomIntField:
```

Comment:
What is this? Did we provide hardcoded values for custom attributes that we use in our own unit tests?
No, it shouldn't be here. Our tests should register these attributes in the test environment. Like application developers would do it when they write tests for their workflows that could use absolutely any custom attributes. So, we have to provide a `registerSearchAttribute()` or something like that in our test framework.